File: bus/sdk/go/caller_test.go
// SPDX-License-Identifier: MIT

package sessionkit

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	"github.com/antst/sessionbus/bus/sdk/go/internal/rpc"
	"github.com/antst/sessionbus/bus/sdk/go/protocol"
	"github.com/antst/sessionbus/bus/sdk/go/testsocket"
)

func TestCallerMapsWireMethods(t *testing.T) {
	var seen []struct {
		method string
		params any
	}
	call := func(_ context.Context, method string, params, result any) error {
		seen = append(seen, struct {
			method string
			params any
		}{method, params})
		results := map[string]any{
			"session.list":   SessionListResult{Sessions: []SessionSummary{}},
			"message.send":   MessageSendResult{MessageID: "message", Deliveries: []protocol.MessageSendDelivery{}},
			"lane.describe":  LaneDescribeResult{Product: "example-peer", SupportedOpenFields: []string{}, ExtraArguments: []ExtraArgument{}},
			"lane.spawn":     LaneSpawnResult{SessionID: "lane@local"},
			"turn.run":       RunStatus{SessionID: "lane@local", RunID: "g/1", State: "done", Result: &TurnResult{Outcome: "completed", Result: "done"}},
			"turn.interrupt": struct{}{},
			"session.close":  struct{}{},
		}
		raw, _ := json.Marshal(results[method])
		return json.Unmarshal(raw, result)
	}
	c := newCaller(call)
	ctx := context.Background()
	table := []struct {
		method string
		params any
		call   func() error
	}{
		{"session.list", SessionListRequest{}, func() error { _, err := c.List(ctx, SessionListRequest{}); return err }},
		{"message.send", MessageSendRequest{Target: "lane", Message: "hello"}, func() error { _, err := c.Send(ctx, MessageSendRequest{Target: "lane", Message: "hello"}); return err }},
		{"lane.describe", LaneDescribeRequest{Product: "example-peer"}, func() error { _, err := c.Describe(ctx, LaneDescribeRequest{Product: "example-peer"}); return err }},
		{"lane.spawn", LaneSpawnRequest{Name: "child", Product: "example-peer", Open: &OpenOptions{}}, func() error {
			_, err := c.Spawn(ctx, LaneSpawnRequest{Name: "child", Product: "example-peer", Open: &OpenOptions{}})
			return err
		}},
		{"lane.spawn", LaneSpawnRequest{ResumeSessionID: "lane@local"}, func() error { _, err := c.Resume(ctx, "lane@local"); return err }},
		{"turn.run", TurnRunRequest{SessionID: "lane@local", Input: "work"}, func() error { _, err := c.Run(ctx, TurnRunRequest{SessionID: "lane@local", Input: "work"}); return err }},
		{"turn.interrupt", SessionTarget{SessionID: "lane@local"}, func() error { return c.Interrupt(ctx, SessionTarget{SessionID: "lane@local"}) }},
		{"session.close", SessionCloseRequest{SessionID: "lane@local", Forget: true}, func() error { return c.Close(ctx, SessionCloseRequest{SessionID: "lane@local", Forget: true}) }},
	}
	for index, test := range table {
		if err := test.call(); err != nil || seen[index].method != test.method || !reflect.DeepEqual(seen[index].params, test.params) {
			t.Fatalf("call %d: got %#v, err %v", index, seen[index], err)
		}
	}
}

func TestCallerSchemaErrorNamesPathAndConstraint(t *testing.T) {
	called := false
	caller := NewCaller(func(context.Context, string, any) (json.RawMessage, error) {
		called = true
		return nil, nil
	})
	_, err := caller.Spawn(context.Background(), LaneSpawnRequest{Name: "child", ResumeSessionID: "lane@local"})
	want := `LaneSpawnRequest: "name" is not allowed with "resume_session_id"`
	if err == nil || err.Error() != want || called {
		t.Fatalf("error = %v, called = %t", err, called)
	}
}

func TestDialIsOneShotFramedClient(t *testing.T) {
	root := testsocket.Directory(t)
	t.Setenv("XDG_RUNTIME_DIR", root)
	t.Setenv("SESSIONBUS_SOCKET", "")
	path := filepath.Join(root, "sessionbus", "presence.sock")
	if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
		t.Fatal(err)
	}
	if got := Socket(); got != path {
		t.Fatalf("socket = %q, want %q", got, path)
	}
	listener, err := net.Listen("unix", path)
	if err != nil {
		t.Fatal(err)
	}
	defer listener.Close()
	go func() {
		fd, _ := listener.Accept()
		requests := make(chan *rpc.Request, 1)
		server := rpc.New(fd, false, func(_ context.Context, request *rpc.Request) { requests <- request })
		_ = server.Result(<-requests, SessionListResult{Sessions: []SessionSummary{}})
	}()
	client, err := Dial("")
	if err != nil {
		t.Fatal(err)
	}
	caller := NewCaller(client.Call)
	listed, err := caller.List(context.Background(), SessionListRequest{})
	if err != nil || len(listed.Sessions) != 0 {
		t.Fatalf("result = %#v, err %v", listed, err)
	}
	if err = client.Close(); err != nil {
		t.Fatal(err)
	}
}

func assertJSON(t *testing.T, value, want any) {
	t.Helper()
	raw, err := json.Marshal(value)
	expected, wantErr := json.Marshal(want)
	if err != nil || wantErr != nil || string(raw) != string(expected) {
		t.Fatalf("json = %s, want %s, errors %v / %v", raw, expected, err, wantErr)
	}
}

func isCode(err error, code int) bool {
	var value *ProtocolError
	return errors.As(err, &value) && value.Code == code
}

type callerFixtures struct {
	Sequences struct {
		CrossedRehello crossedRehelloFixture `json:"crossed_rehello"`
	}
}

func loadCallerFixtures(t *testing.T) callerFixtures {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("protocol", "caller-sugar.fixtures.json"))
	if err != nil {
		t.Fatal(err)
	}
	var fixtures callerFixtures
	if err = json.Unmarshal(raw, &fixtures); err != nil {
		t.Fatal(err)
	}
	return fixtures
}

func TestCallerCancelledWaitRetainsWorkerResult(t *testing.T) {
	for _, action := range []bool{false, true} {
		for _, bounded := range []bool{false, true} {
			for _, collect := range []string{"status", "wait"} {
				t.Run(fmt.Sprintf("action=%t/bounded=%t/collect=%s", action, bounded, collect), func(t *testing.T) {
					release, entered := make(chan struct{}), make(chan struct{})
					requests := make(chan *rpc.Request, 16)
					p := &cursorProduct{run: func(context.Context, *Run, RunInput) (TurnResult, error) {
						close(entered)
						<-release
						return TurnResult{Outcome: "completed", Result: "retained"}, nil
					}}
					_, wire := cursorWire(t, p, func(r *rpc.Request) { requests <- r })
					cursorOpen(t, wire)
					if err := cursorExecute(wire, "native@local", "g/1"); err != nil {
						t.Fatal(err)
					}
					<-entered
					caller := newCaller(wire.Call)
					request := WaitRequest{SessionID: "native@local", RunID: "g/1"}
					if bounded {
						timeout := int64(60000)
						request.TimeoutMS = &timeout
					}
					ctx, cancel := context.WithCancel(context.Background())
					defer cancel()
					waiting := make(chan error, 1)
					go func() {
						if action {
							raw, _ := json.Marshal(request)
							_, err := caller.Action(ctx, "wait", raw)
							waiting <- err
						} else {
							_, err := caller.WaitContext(ctx, request)
							waiting <- err
						}
					}()
					for r := range requests {
						if r.Method == "turn.wait" {
							break
						}
					}
					cancel()
					if err := <-waiting; !errors.Is(err, context.Canceled) {
						t.Fatalf("cancelled wait: %v", err)
					}
					close(release)
					// A replacement Caller reaches the same resident cursor; the canceled
					// correlation drains its late response before this subsequent reply.
					replacement := newCaller(wire.Call)
					got, err := replacement.Wait(WaitRequest{SessionID: request.SessionID, RunID: request.RunID})
					if err != nil || got.Result == nil || got.Result.Result != "retained" {
						t.Fatalf("retained %#v %v", got, err)
					}
					if _, err = caller.WaitContext(ctx, request); !errors.Is(err, context.Canceled) {
						t.Fatalf("pre-cancel wait %v", err)
					}
					if collect == "status" {
						got, err = replacement.Status(context.Background(), ReadRequest{SessionID: request.SessionID, RunID: request.RunID})
					} else {
						got, err = replacement.Wait(request)
					}
					if err != nil || got.Result == nil || got.Result.Result != "retained" {
						t.Fatalf("second read %#v %v", got, err)
					}
					ref := RunRef{SessionID: request.SessionID, RunID: request.RunID}
					if err = replacement.Ack(ctx, ref); !errors.Is(err, context.Canceled) {
						t.Fatalf("pre-canceled ack %v", err)
					}
					if _, err = replacement.Status(context.Background(), ReadRequest{SessionID: ref.SessionID, RunID: ref.RunID}); err != nil {
						t.Fatal(err)
					}
					if err = replacement.Ack(context.Background(), ref); err != nil {
						t.Fatal(err)
					}
					_, err = replacement.Status(context.Background(), ReadRequest{SessionID: ref.SessionID, RunID: ref.RunID})
					cursorCode(t, err, protocol.UnknownSession)
				})
			}
		}
	}
}

func TestCallerSubmittedAckSettlesDespiteCancellation(t *testing.T) {
	client, server := net.Pipe()
	requests := make(chan *rpc.Request, 1)
	daemon := rpc.New(server, false, func(_ context.Context, r *rpc.Request) { requests <- r })
	wire := rpc.New(client, true, func(context.Context, *rpc.Request) {})
	t.Cleanup(func() { _ = wire.Close(); _ = daemon.Close() })
	caller := newCaller(wire.Call)
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	done := make(chan error, 1)
	go func() { done <- caller.Ack(ctx, RunRef{SessionID: "native@local", RunID: "g/1"}) }()
	request := <-requests
	cancel()
	if err := daemon.Result(request, struct{}{}); err != nil {
		t.Fatal(err)
	}
	if err := <-done; err != nil {
		t.Fatalf("submitted ack abandoned: %v", err)
	}
}

File: bus/sdk/js/caller.test.js
// SPDX-License-Identifier: MIT

"use strict";

const assert = require("node:assert/strict");
const fs = require("node:fs");
const path = require("node:path");
const test = require("node:test");
const { ACTIONS, Caller, connectPeer, Connection, ProtocolError } = require("./index.js");
const { pair, deferred } = require("./test-support.js");

const fixtures = JSON.parse(fs.readFileSync(path.join(__dirname, "../go/protocol/caller-sugar.fixtures.json"), "utf8"));

const replies = {
  "session.list": { sessions: [] },
  "message.send": { message_id: "message", deliveries: [] },
  "lane.describe": { product: "example-peer", supported_open_fields: [], extra_arguments: [] },
  "lane.spawn": { session_id: "lane@local" },
  "turn.run": { session_id: "lane@local", run_id: "g/1", state: "done", result: { outcome: "completed", result: "done" } },
  "turn.interrupt": {},
  "session.close": {},
};

test("caller maps operations to closed wire requests", async (t) => {
  const [clientSocket, daemonSocket] = pair();
  const seen = [];
  const daemon = new Connection(daemonSocket, false, (request) => { seen.push([request.method, request.params]); void daemon.result(request, replies[request.method]); });
  const connection = new Connection(clientSocket, true);
  const caller = new Caller(connection);
  t.after(() => { connection.close(); daemon.close(); });

  const table = [
    [() => caller.list({}), "session.list", {}],
    [() => caller.send({ target: "lane", message: "hello" }), "message.send", { target: "lane", message: "hello" }],
    [() => caller.describe({ product: "example-peer" }), "lane.describe", { product: "example-peer" }],
    [() => caller.spawn({ name: "child", product: "example-peer", open: {} }), "lane.spawn", { name: "child", product: "example-peer", open: {} }],
    [() => caller.resume("lane@local"), "lane.spawn", { resume_session_id: "lane@local" }],
    [() => caller.run({ session_id: "lane@local", input: "work" }), "turn.run", { session_id: "lane@local", input: "work" }],
    [() => caller.interrupt({ session_id: "lane@local" }), "turn.interrupt", { session_id: "lane@local" }],
    [() => caller.close({ session_id: "lane@local", forget: true }), "session.close", { session_id: "lane@local", forget: true }],
  ];
  for (const [call, method, params] of table) { await call(); assert.deepEqual(seen.shift(), [method, params]); }

  const actions = [
    ["list", {}, "session.list", {}],
    ["send", { target: "lane", message: "hello" }, "message.send", { target: "lane", message: "hello" }],
    ["spawn", { name: "child", product: "example-peer", open: {} }, "lane.spawn", { name: "child", product: "example-peer", open: {} }],
    ["describe", { product: "example-peer" }, "lane.describe", { product: "example-peer" }],
    ["run", { session_id: "lane@local", input: "work" }, "turn.run", { session_id: "lane@local", input: "work" }],
    ["interrupt", { session_id: "lane@local" }, "turn.interrupt", { session_id: "lane@local" }],
    ["close", { session_id: "lane@local" }, "session.close", { session_id: "lane@local" }],
    ["forget", { session_id: "lane@local" }, "session.close", { session_id: "lane@local", forget: true }],
  ];
  for (const [action, args, method, params] of actions) { await caller.action(action, args); assert.deepEqual(seen.shift(), [method, params]); }
  await assert.rejects(caller.spawn({ resume_session_id: "lane@local", name: "child" }), (error) => error.message === `LaneSpawnRequest: "name" is not allowed with "resume_session_id"`);
  assert.equal(seen.length, 0);
  assert.deepEqual(ACTIONS, ["list", "send", "spawn", "describe", "run", "start", "wait", "status", "interrupt", "close", "forget", "ack"]);
  await assert.rejects(caller.action("unknown", {}), /unknown action/);
  await assert.rejects(caller.action("list", { extra: true }), /SessionListRequest: "extra" is not allowed/);
  await assert.rejects(caller.action("status", { turn_id: "missing", extra: true }), /ReadRequest:/);
});

test("caller operations use shared wire fixtures", async (t) => {
 const [client, server] = pair(); let current;
 const daemon = new Connection(server, false, (request) => { assert.equal(request.method, current.method); assert.deepEqual(request.params, current.request); void daemon.result(request, current.result); });
 const connection = new Connection(client, true), caller = new Caller(connection);
 t.after(() => { connection.close(); daemon.close(); });
 for (const row of fixtures.operations) { current = row; assert.deepEqual(await caller.action(row.action, row.request), row.result); }
});

test("submitted ack settles after cancellation; pre-aborted ack sends nothing", async (t) => {
 const [client, server] = pair(), received = deferred(); let writes = 0;
 const daemon = new Connection(server, false, (request) => { writes++; received.resolve(request); });
 const connection = new Connection(client,true), caller = new Caller(connection);
 t.after(() => { connection.close(); daemon.close(); });
 const ref = { session_id: "lane@local", run_id: "g/1" }, cancel = new AbortController();
 const ack = caller.action("ack",ref,cancel.signal), request = await received.promise;
 cancel.abort(new Error("cancelled")); await daemon.result(request,{}); assert.deepEqual(await ack,{});
 await assert.rejects(caller.ack(ref,cancel.signal),/cancelled/); assert.equal(writes,1);
});

test("crossed rehello preserves the fixture's newest identity", async (t) => {
  const [clientSocket, daemonSocket] = pair();
  const queued = [], waiting = [];
  const next = () => queued.length ? Promise.resolve(queued.shift()) : new Promise((resolve) => waiting.push(resolve));
  const daemon = new Connection(daemonSocket, false, (request) => { const resolve = waiting.shift(); if (resolve) resolve(request); else queued.push(request); });
  const row = structuredClone(fixtures.sequences.crossed_rehello);
  const env = { SESSIONBUS_SOCKET: "/fixture/socket", SESSIONBUS_LOCAL_KEY: "" };
  const peer = connectPeer(row.identity, async () => ({ disposition: "injected" }), env, { connect: () => clientSocket, schedule: () => {} });
  t.after(() => { peer.shutdown(); daemon.close(); });

  row.identity.info.nested.value = "mutated";
  const initial = await next();
  assert.deepEqual([initial.params.name, initial.params.info.nested.value], ["initial", "initial"]);
  await daemon.result(initial, {}); await peer.ready;

  const first = peer.rehello(undefined, row.first.name, row.first.info); const firstRequest = await next(); row.first.info.nested.value = "mutated";
  const second = peer.rehello(undefined, row.second.name, row.second.info); const secondRequest = await next(); row.second.info.nested.value = "mutated";
  await daemon.result(secondRequest, {}); await second;
  await daemon.result(firstRequest, {});
  const corrective = await next();
  assert.deepEqual([corrective.params.name, corrective.params.info.nested.value], ["second", "second"]);
  await daemon.result(corrective, {}); await first;
  assert.deepEqual([peer.identity.name, peer.identity.info.nested.value], ["second", "second"]);
});

